### PR TITLE
Use generator for creating easily distributable Go packages

### PR DESCRIPTION
**Pull request by afiune**
## Description
Since we are starting to use Go inside our CLI world at Chef, we need a way to
make all new code distributable, this is a proposal that shows a pattern we can
use to generate distributable Go packages.

The whole idea is detailed at https://github.com/afiune/godist#godist and this
is just an example of how easy would it be to leverage this automation.

TODO's:

- [x] (https://github.com/chef/go-libs/pull/12) Move https://github.com/afiune/godist to https://github.com/chef/go-libs 
- [x] (https://github.com/chef/go-libs/pull/14) Add a few more dist variables like `.chef` directory and `Chef` the company name

## Related Issue

Closes https://github.com/chef/chef-analyze/issues/63

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: vendor/github.com/afiune/godist/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+# godist
+A simple generator for creating easily distributable Go packages
+
+This automation is providing an easy way to generate variables that can be configured,
```
Comment (mjingle) [inline vendor/github.com/afiune/godist/README.md]:
```suggestion
This automation provides an easy way to generate variables that can be configured,
```

File: vendor/github.com/afiune/godist/README.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+globally across multiple go packages, things like trademarks, product names, websites,
```
Comment (mjingle) [inline vendor/github.com/afiune/godist/README.md]:
```suggestion
globally across multiple go packages, such as trademarks, product names, websites,
```

File: vendor/github.com/afiune/godist/README.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+etc., this generator should be defined as a `go:generate` comment and run at build time
```
Comment (mjingle) [inline vendor/github.com/afiune/godist/README.md]:
```suggestion
etc. This generator should be defined as a `go:generate` comment and run at build time,
```

File: vendor/github.com/afiune/godist/README.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
+# godist
+A simple generator for creating easily distributable Go packages
```
Comment (mjingle) [inline vendor/github.com/afiune/godist/README.md]:
```suggestion

A simple generator for creating easily distributable Go packages
```

File: vendor/github.com/afiune/godist/README.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+This automation is providing an easy way to generate variables that can be configured,
+globally across multiple go packages, things like trademarks, product names, websites,
+etc., this generator should be defined as a `go:generate` comment and run at build time
+using the `go generate` command.
+
+### Example-1
+Simple use within a single main package, create a file called dist_gen.go with:
```
Comment (mjingle) [inline vendor/github.com/afiune/godist/README.md]:
```suggestion
Our first example involves a simple use within a single main package. First, create a file called `dist_gen.go` with the following command:
```

File: vendor/github.com/afiune/godist/README.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+```go
+package main
+//go:generate go run github.com/afiune/godist
+```
+
+The automation will deploy a file called `dist.go` with all the variables defined inside
+the JSON file `glob_dist.json` inside this repository. (find a real example at
```
Comment (mjingle) [inline vendor/github.com/afiune/godist/README.md]:
```suggestion
the JSON file `glob_dist.json` inside this repository. (See a real example at
```

File: vendor/github.com/afiune/godist/README.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+[example-main/](example-main))
```
Comment (mjingle) [inline vendor/github.com/afiune/godist/README.md]:
```suggestion
[example-main/](example-main).)
```

File: vendor/github.com/afiune/godist/README.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+package dist
+//go:generate go run github.com/afiune/godist global.go dist
+```
+
+This usage is for go projects that has multiple packages, by creating a single `dist` package
+inside your repository, you can import the generated package in any other packages. (find a
+real example at [example-multi-pkg/](example-multi-pkg))
+
+### Example-3
+Using a custom JSON file containing global variables
```
Comment (mjingle) [inline vendor/github.com/afiune/godist/README.md]: Are you running another command with this third example? This line seems to need more context.

File: vendor/github.com/afiune/godist/README.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+```go
+package dist
+//go:generate go run github.com/afiune/godist global.go dist http://shorturl.at/arxQX
+```
+
+To fully customize this automation, a user can also provide a custom URL poiting to a JSON file
```
Comment (mjingle) [inline vendor/github.com/afiune/godist/README.md]:
```suggestion
To fully customize this automation, a user can also provide a custom URL pointing to a JSON file
```

File: vendor/github.com/afiune/godist/README.md [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+### Example-2
+Multi-package, create a go package called dist/ with a file called `gen.go` with:
+
+```go
+package dist
+//go:generate go run github.com/afiune/godist global.go dist
+```
+
+This usage is for go projects that has multiple packages, by creating a single `dist` package
```
Comment (mjingle) [inline vendor/github.com/afiune/godist/README.md]:
```suggestion
This usage is for go projects that has multiple packages. By creating a single `dist` package
```

File: vendor/github.com/afiune/godist/README.md [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+inside your repository, you can import the generated package in any other packages. (find a
```
Comment (mjingle) [inline vendor/github.com/afiune/godist/README.md]:
```suggestion
inside your repository, you can import the generated package in any other packages. (See a
```

File: vendor/github.com/afiune/godist/README.md [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+real example at [example-multi-pkg/](example-multi-pkg))
```
Comment (mjingle) [inline vendor/github.com/afiune/godist/README.md]:
```suggestion
real example at [example-multi-pkg/](example-multi-pkg).)
```

File: vendor/github.com/afiune/godist/README.md [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+package main
+//go:generate go run github.com/afiune/godist
+```
+
+The automation will deploy a file called `dist.go` with all the variables defined inside
+the JSON file `glob_dist.json` inside this repository. (find a real example at
+[example-main/](example-main))
+
+### Example-2
+Multi-package, create a go package called dist/ with a file called `gen.go` with:
```
Comment (mjingle) [inline vendor/github.com/afiune/godist/README.md]:
```suggestion
In our second example involving multi-package. Create a go package called dist/ with a file called `gen.go` with the following command:
```

**Comment by afiune** [comment]
I'm coming back to this work to rebase it and get it reviewed so we can merge it!

tenor-71239725

File: cmd/report.go [hunk 14]
```diff
@@ -211,7 +212,7 @@ func init() {
 	reportCmd.PersistentFlags().StringVarP(
 		&reportsFlags.credsFile,
 		"credentials", "c", "",
-		"Chef credentials file (default $HOME/.chef/credentials)",
+		"Chef credentials file (default $HOME/.chef/credentials)", // @afiune add Chef and .chef to dist vars
```
Comment (tyler-ball) [inline cmd/report.go]: This still looks like a valid TODO

Comment (tyler-ball) [inline cmd/report.go]: This still looks like a valid TODO

File: vendor/github.com/afiune/godist/tiny_glob_dist.json [hunk 15]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (tyler-ball) [inline vendor/github.com/afiune/godist/tiny_glob_dist.json]: Do we need to have this file present?

**Comment by tyler-ball** [comment]
Could we potentially update the deps in another no-review-needed PR and get those merged to master so this PR isn't so cluttered?

**Comment by afiune** [comment]
### We are ready to merge this work!

tenor-173851513

**Pull request merged**